### ISSUE TITLE
Minimization with frozen particles

### ISSUE THREAD

**Issue by SHervoe-Hansen** [openmm/openmm#3884]
Hi everyone,

Recently I have been looking into how one can minimize systems containing frozen particles. In the past, I have attempted to set the mass to zero for the particles that are to be frozen, which works during the normal integration:
```python
for atom in protein_atoms:
    system.setParticleMass(atom.index, 0.000*u.dalton)
integrator = mm.LangevinMiddleIntegrator(298.15*u.kelvin, 1.0/u.picoseconds, 2.0*u.femtoseconds)
sim = app.Simulation(pdb.topology, system, integrator, platform, properties)
sim.context.setPositions(pdb.positions)
sim.minimizeEnergy()
```
However, it is my impression that all particles are still being minimized using this approach. Instead, I was thinking if I should use the [setIntegrationForceGroups](http://docs.openmm.org/latest/api-python/generated/openmm.openmm.LangevinMiddleIntegrator.html#openmm.openmm.LangevinMiddleIntegrator.setIntegrationForceGroups) method from the integrator. When calling [getIntegrationForceGroups](http://docs.openmm.org/latest/api-python/generated/openmm.openmm.LangevinMiddleIntegrator.html#openmm.openmm.LangevinMiddleIntegrator.getIntegrationForceGroups)(self) it returns -1, which I assume means that the integrator is acting on all particles? 

To summarize my questions:
a. Should the ForceGroups not change for the integrator when the system contains particles with zero mass?
b. Could you provide a minimal example of how to freeze a set of particles during minimization? If done by changing the ForceGroups please show one can use either set or int as explained in the [documentation](http://docs.openmm.org/latest/api-python/generated/openmm.openmm.LangevinMiddleIntegrator.html#openmm.openmm.LangevinMiddleIntegrator.setIntegrationForceGroups).

**Comment by peastman** [comment]
You don't need to do anything.  The minimizer won't move massless particles.

**Comment by SHervoe-Hansen** [comment]
Hi Peter,
That is also my understanding, however, then I don't understand the following:
If I take a protein structure from a molecular dynamics simulation I have previously conducted and place it in a vacuum, then the energy of the initial configuration before and after minimization is different even when all particles in the system possess 0 mass. To check the structure I am also computing the RMSD between the initial configuration (input for openMM), and the protein structure before minimization and after minimization. Again there is no difference before the minimization, while there is a difference after minimization.

I have attached a tarball ([test.tar.gz](https://github.com/openmm/openmm/files/10323365/test.tar.gz)) containing an example that can reproduce the above with the python script depending on python3, openMM, and mdtraj.

I am sorry in advance if turns out to be me having overlooked some small little detail.

**Comment by SHervoe-Hansen** [comment]
So I did a bit more testing and it turns out to be the constraints.
I had the following line in my code:
```python
system = forcefield.createSystem(pdb.topology, nonbondedMethod=app.PME, ewaldErrorTolerance=0.0005,
 switchDistance=1*u.nanometer, nonbondedCutoff=1.2*u.nanometers, constraints=app.HBonds)
```
Setting the constraints to `None` fixes everything. Is it possible to enable the constraints following the minimization in an easy way without reinitializing the system?

**Comment by peastman** [comment]
I think I see the problem.  The minimizer doesn't know how to handle constraints directly.  So we replace them with harmonic restraints, and successively tighten them until minimization produces positions that satisfy all constraints to the required tolerance.  With the regular forces, we set the force on any massless particle to zero.  But we don't do it for the restraint forces.  That's easy to fix.
